rsl/rsl.go

```diff
@@ -39,37 +43,132 @@ func InitializeNamespace() error {
+func NewRevocationEntry(refName string, commitID plumbing.Hash, revokedEntryIDs []plumbing.Hash) *RSLEntry {
+	return &RSLEntry{RefName: refName, CommitID: commitID, Revoke: revokedEntryIDs}
+}
+
+// Commit creates a commit object in the RSL for the RSLEntry.
+// If the entry is a revocation entry, Commit verifies that its ref matches that
+// of the entry it is revoking.
+func (r *RSLEntry) Commit(sign bool) error {
 	repo, err := common.GetRepositoryHandler()
 	if err != nil {
 		return err
 	}
-	message := fmt.Sprintf("%s: %s", refName, commitID.String())
 
-	return gitinterface.Commit(repo, plumbing.ZeroHash, RSLRef, message, sign)
+	for _, revoked := range r.Revoke {
+		revokedEntry, err := GetEntry(revoked)
+		if err != nil {
+			return err
+		}
+
+		if revokedEntry.RefName != r.RefName {
+			return ErrRSLRevokeEntryHasInvalidRef
+		}
+	}
+
+	return gitinterface.Commit(repo, plumbing.ZeroHash, RSLRef, r.createCommitMessage(), sign)
+}
+
+func (r *RSLEntry) createCommitMessage() string {
+	lines := []string{
+		fmt.Sprintf("ref: %s", r.RefName),
+		fmt.Sprintf("commit: %s", r.CommitID.String()),
+	}
+
+	for _, revoked := range r.Revoke {
+		lines = append(lines, fmt.Sprintf("revoke: %s", revoked.String()))
+	}
+	return strings.Join(lines, "\n")
 }
 
 // GetLatestEntry returns the latest entry available locally in the RSL.
 // TODO: There is no information yet about the signature for the entry.
-func GetLatestEntry() (string, plumbing.Hash, error) {
+func GetLatestEntry() (*RSLEntry, error) {
 	repo, err := common.GetRepositoryHandler()
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
 	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
 	commitObj, err := repo.CommitObject(ref.Hash())
 	if err != nil {
-		return "", plumbing.ZeroHash, err
+		return &RSLEntry{}, err
 	}
 
-	messageSplit := strings.Split(commitObj.Message, ":")
+	return parseRSLEntryMessage(commitObj.Message), nil
+}
 
-	return strings.Trim(messageSplit[0], " "),
-		plumbing.NewHash(strings.Trim(messageSplit[1], " ")), nil
+// GetEntry returns the entry corresponding to entryID.
+// TODO: There is no information yet about the signature for the entry.
+func GetEntry(entryID plumbing.Hash) (*RSLEntry, error) {
+	repo, err := common.GetRepositoryHandler()
+	if err != nil {
+		return &RSLEntry{}, err
+	}
+
+	ref, err := repo.Reference(plumbing.ReferenceName(RSLRef), true)
+	if err != nil {
+		return &RSLEntry{}, err
+	}
+
+	iteratorHash := ref.Hash()
+	for {
+		commitObj, err := repo.CommitObject(iteratorHash)
+		if err != nil {
+			return &RSLEntry{}, err
+		}
+
+		if iteratorHash == entryID {
+			return parseRSLEntryMessage(commitObj.Message), nil
+		}
+
+		if len(commitObj.ParentHashes) == 0 {
+			return &RSLEntry{}, ErrRSLEntryNotFound
+		}
+
+		if len(commitObj.ParentHashes) > 1 {
+			return &RSLEntry{}, ErrRSLBranchDetected
```

Comment:
It'd be interesting to see how we can handle this 🤔 

Merge commits are not uncommon given it's the default for GitHub PRs.

(fine with ignoring for now though)

Reply:
My thinking is we can control this as all RSL entries are created by gittuf. We only care about merge commits in `refs/gittuf/reference-state-log` to avoid any forking in the log.